source/s3_auto_ranged_put.c

```diff
@@ -332,88 +271,44 @@ static int s_s3_auto_ranged_put_next_request(
 static int s_s3_auto_ranged_put_prepare_request(
     struct aws_s3_meta_request *meta_request,
     struct aws_s3_client *client,
-    struct aws_s3_request *request) {
+    struct aws_s3_vip_connection *vip_connection,
+    bool is_initial_prepare) {
     AWS_PRECONDITION(meta_request);
     AWS_PRECONDITION(client);
+    (void)client;
+    (void)is_initial_prepare;
+
+    struct aws_s3_request *request = vip_connection->request;
     AWS_PRECONDITION(request);
 
     struct aws_s3_auto_ranged_put *auto_ranged_put = meta_request->impl;
     AWS_PRECONDITION(auto_ranged_put);
 
     struct aws_http_message *message = NULL;
-    struct aws_s3_part_buffer *part_buffer = NULL;
-
-    /* If we're not sending the entire object, then we need to grab a part buffer */
-    if (request->desc_data.request_tag != AWS_S3_AUTO_RANGED_PUT_REQUEST_TAG_ENTIRE_OBJECT) {
-
-        part_buffer = aws_s3_client_get_part_buffer(client, request->desc_data.part_number);
-
-        if (part_buffer == NULL) {
-            AWS_LOGF_ERROR(
-                AWS_LS_S3_META_REQUEST,
-                "id=%p Could not get part buffer for request with tag %d for auto-ranged-get meta request.",
-                (void *)meta_request,
-                request->desc_data.request_tag);
-
-            aws_raise_error(AWS_ERROR_S3_NO_PART_BUFFER);
-            goto part_buffer_get_failed;
-        }
-    }
 
-    switch (request->desc_data.request_tag) {
-
-        /* If we're grabbing the whole object, just use the original message. */
-        case AWS_S3_AUTO_RANGED_PUT_REQUEST_TAG_ENTIRE_OBJECT: {
-            message =
-                aws_s3_message_util_copy_http_message(meta_request->allocator, meta_request->initial_request_message);
-
-            AWS_FATAL_ASSERT(message);
-            break;
-        }
+    switch (request->request_tag) {
 
         case AWS_S3_AUTO_RANGED_PUT_REQUEST_TAG_PART: {
 
             bool error_occurred = false;
 
-            s_s3_auto_ranged_put_lock_synced_data(auto_ranged_put);
+            if (is_initial_prepare) {
 
-            uint64_t range_start = (request->desc_data.part_number - 1) * meta_request->part_size;
-            struct aws_input_stream *initial_body_stream = meta_request->synced_data.initial_body_stream;
+                AWS_ASSERT(request->request_body.capacity == 0);
```

Comment:
does `request_body.capacity == 0` universally indicate that this is the initial attempt? If so, we could do away with the `is_initial_prepare` flag, right?

Reply:
It's what I originally had, but it felt a bit hacky to me, because you're having to intuit if this is the first prepare.  (Not everything will need to set up a request body either.)